### PR TITLE
added tests for repositories, services and controllers of book and ca…

### PR DESCRIPTION
**Pull request by VitaliiPotseluiko**
Screen shots 
image

### CODE REVIEW

File: src/test/java/com/springboot/intro/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+        MvcResult result = mockMvc.perform(get("/api/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        BookResponseDto[] actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                BookResponseDto[].class);
+
+        assertEquals(expected.size(), actual.length);
+        assertEquals(expected, Arrays.stream(actual).toList());
```
Comment (karma-o) [inline src/test/java/com/springboot/intro/controller/BookControllerTest.java:234]:
```suggestion
        assertEquals(expected, List.of(actual));
```

File: src/test/java/com/springboot/intro/controller/CategoryControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+        MvcResult result = mockMvc.perform(put("/api/categories/2")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+                .andReturn();
+        ErrorValidationDto actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                ErrorValidationDto.class);
+        assertEquals("name can't be blank", actual.getErrors()[0]);
+        assertEquals(HttpStatus.BAD_REQUEST, actual.getStatus());
```
Comment (karma-o) [inline src/test/java/com/springboot/intro/controller/CategoryControllerTest.java:196]: let's separate our code into 3 sections split with a space:
* **Given** - essentially setting up our `expected` variables and conditions like `mock.when`
* **When** - performing actions that we are testing and getting `actual` variable
* **Then** - assertion of the given and actual values

And try to keep that separation across all tests

Comment (karma-o) [reply src/test/java/com/springboot/intro/controller/CategoryControllerTest.java:196]: more about it [here](https://martinfowler.com/bliki/GivenWhenThen.html)

File: src/test/java/com/springboot/intro/controller/CategoryControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNotFound())
+                .andReturn();
+        StatusErrorDto actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                StatusErrorDto.class);
+        assertEquals("There is no such category by id 4", actual.getError());
+        assertEquals(HttpStatus.NOT_FOUND, actual.getStatus());
+    }
+}
```
Comment (karma-o) [inline src/test/java/com/springboot/intro/controller/CategoryControllerTest.java:221]:
```suggestion
}

```

File: src/test/java/com/springboot/intro/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class BookControllerTest {
```
Comment (karma-o) [inline src/test/java/com/springboot/intro/controller/BookControllerTest.java:44]: I know that we have a requirement of 50% coverage is enough, but we still have search and delete endpoints not tested, which should be done on the real project, but we can skip them here

File: src/test/java/com/springboot/intro/controller/CategoryControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class CategoryControllerTest {
```
Comment (karma-o) [inline src/test/java/com/springboot/intro/controller/CategoryControllerTest.java:42]: same here, basically, each method and endpoint in every controller\service\util class should be tested on the real project, and the whole project should have coverage of around 80-90% of methods total. Just some info for the future here :)

File: src/test/java/com/springboot/intro/controller/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    @DisplayName("""
+            save invalid book
+            """)
+    public void save_InvalidRequestDto_BadRequest() throws Exception {
+        BookRequestDto invalidRequest = new BookRequestDto();
+        invalidRequest.setTitle("Title6").setAuthor(null).setIsbn("2400000032632")
+                .setPrice(BigDecimal.valueOf(-1)).setCoverImage("cow6").setDescription("d6")
+                .setCategoryId(2L);
+        String jsonRequest = objectMapper.writeValueAsString(invalidRequest);
+        List<String> errors = List.of("author can't be blank", "price must be bigger than 0");
```
[the remark this answers is not in the document]
Comment (VitaliiPotseluiko) [reply src/test/java/com/springboot/intro/controller/BookControllerTest.java:117]: ++

File: src/test/java/com/springboot/intro/controller/CategoryControllerTest.java (hunk 2 above)
Comment (VitaliiPotseluiko) [reply src/test/java/com/springboot/intro/controller/CategoryControllerTest.java:196]: ++

File: src/test/java/com/springboot/intro/controller/BookControllerTest.java (hunk 4 above)
Comment (VitaliiPotseluiko) [reply src/test/java/com/springboot/intro/controller/BookControllerTest.java:44]: Ok
++

File: src/test/java/com/springboot/intro/controller/CategoryControllerTest.java (hunk 5 above)
Comment (VitaliiPotseluiko) [reply src/test/java/com/springboot/intro/controller/CategoryControllerTest.java:42]: ++

**Pull request merged**